packages/instantsearch.js/src/middlewares/createInsightsMiddleware.ts

```diff
@@ -241,21 +254,64 @@ export function createInsightsMiddleware<
           setUserTokenToSearch(anonymousUserToken, true);
         }
 
-        // We consider the `userToken` coming from a `init` call to have a higher
-        // importance than the one coming from the queue.
-        if (userTokenBeforeInit) {
-          setUserTokenToSearch(userTokenBeforeInit, true);
-          insightsClient('setUserToken', userTokenBeforeInit);
-        } else if (queuedUserToken) {
-          setUserTokenToSearch(queuedUserToken, true);
-          insightsClient('setUserToken', queuedUserToken);
+        function setUserToken(
+          token: string | number,
+          userToken?: string | number,
+          authenticatedUserToken?: string | number
+        ) {
+          setUserTokenToSearch(token, true);
+
+          if (userToken) {
+            insightsClient('setUserToken', userToken);
+          }
+          if (authenticatedUserToken) {
+            insightsClient('setAuthenticatedUserToken', authenticatedUserToken);
+          }
+        }
+
+        // We consider the `userToken` or `authenticatedUserToken` before an
+        // `init` call of higher importance than one from the queue.
+        const tokenBeforeInit =
+          authenticatedUserTokenBeforeInit || userTokenBeforeInit;
+        const queuedToken = queuedAuthenticatedUserToken || queuedUserToken;
+
+        if (tokenBeforeInit) {
+          setUserToken(
+            tokenBeforeInit,
+            userTokenBeforeInit,
+            authenticatedUserTokenBeforeInit
+          );
+        } else if (queuedToken) {
+          setUserToken(
+            queuedToken,
+            queuedUserToken,
+            queuedAuthenticatedUserToken
+          );
         }
 
         // This updates userToken which is set explicitly by `aa('setUserToken', userToken)`
         insightsClient('onUserTokenChange', setUserTokenToSearch, {
           immediate: true,
         });
 
+        // This updates userToken which is set explicitly by `aa('setAuthenticatedtUserToken', authenticatedUserToken)`
+        insightsClient(
+          'onAuthenticatedUserTokenChange',
+          (authenticatedUserToken) => {
+            // If we're unsetting the `authenticatedUserToken`, we revert to the `userToken`
+            if (!authenticatedUserToken) {
+              insightsClient('getUserToken', null, (_, userToken) => {
+                setUserTokenToSearch(userToken);
+              });
+            }
+
+            setUserTokenToSearch(authenticatedUserToken);
```

Comment:
Could there be a race condition here where reverting to the `userToken` would be overridden by an undefined `authenticatedUserToken`?

Reply:
Ah yes, that's it.